### PR TITLE
btop: 1.1.2 (new formula)

### PR DESCRIPTION
**Pull request by nevack**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: Formula/btop.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
+class Btop < Formula
+  desc "Resource monitor. C++ version and continuation of bashtop and bpytop"
+  homepage "https://github.com/aristocratos/btop"
+  url "https://github.com/aristocratos/btop/archive/refs/tags/v1.1.2.tar.gz"
+  sha256 "997376596cc6ca808b5a26f9fe25da274c19c2970aaf757943a009f4c28b758c"
+  license "Apache-2.0"
+
+  depends_on "coreutils" => :build
+  depends_on "gcc" => :build
```
Comment (carlocab) [inline Formula/btop.rb:9]:
```suggestion
  fails_with: :clang

  on_macos do
    depends_on "gcc"
  end
```
Why doesn't Clang work? If you need to use GCC, you'll need to keep it around at runtime.

File: Formula/btop.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+  def install
+    if OS.mac?
+      gcc_major_ver = Formula["gcc"].any_installed_version.major
+      ENV["CXX"] = Formula["gcc"].opt_bin/"g++-#{gcc_major_ver}"
+    end
+
+    system "make", "CXX=#{ENV.cxx}"
```
Comment (carlocab) [inline Formula/btop.rb:17]:
```suggestion
    system "make", "CXX=#{ENV.cxx}"
```

File: Formula/btop.rb (hunk 1 above)
Comment (nevack) [reply Formula/btop.rb:9]: This is what upstream suggests. https://github.com/aristocratos/btop#compilation-osx
I have tried it with clang initially, but `-ftree-loop-vectorize -flto=12 -s` options are not supported.

I should add `STATIC=true` to `make` to "make it work" :)

Thanks for `fails_with :clang` magic!

Comment (nevack) [reply Formula/btop.rb]: Note about `STATIC=true` - upstream suggest to static link to libstdc++/libc++, this is what this flag is for

**Comment by nevack** [comment]
@carlocab It's a bit rough with all that `fails_with gcc: "x"` but everything is working now

File: Formula/btop.rb [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  fails_with :clang # -ftree-loop-vectorize -flto=12 -s
+  # GCC 10 at least is required
+  fails_with gcc: "5"
+  fails_with gcc: "6"
+  fails_with gcc: "7"
+  fails_with gcc: "8"
+  fails_with gcc: "9"
+
+  def install
+    system "make", "CXX=#{ENV.cxx}", "STATIC=true", "STRIP=true"
```
Comment (SMillerDev) [inline Formula/btop.rb:23]: Are we building static libraries here? Because generally we don't want that.

**Review by SMillerDev — DISMISSED** [review]

File: Formula/btop.rb (hunk 3 above)
Comment (carlocab) [reply Formula/btop.rb:23]: I think this is linking statically with `libstdc++`

**Review by carlocab — DISMISSED** [review]
@Bo98 opposed to statically linking `libstdc++` here?

**Comment by Bo98** [comment]
We generally don't statically link against dependencies. Is it necessary here?

**Comment by nevack** [comment]
> We generally don't statically link against dependencies. Is it necessary here?

It is just what upstream suggests to do.

This project requires latest version of gcc to be built, having it installed locally just to run such a small piece of software seems unnecessary.

**Comment by SMillerDev** [comment]
> This project requires latest version of gcc to be built, having it installed locally just to run such a small piece of software seems unnecessary.

The better solution seems fixing the CLang issue, rather than statically linking libstdc++.

**Comment by Bo98** [comment]
Not sure where you see that it is _suggesting_ static (the mention in the README is just documentation of the flag), but it appears this will statically link glibc on Linux, which is something we're unlikely to accept.

**Comment by carlocab** [comment]
Ok, I suggest just make GCC a hard dependency for now and we can fix that when upstream fix compilation with Clang.

**Comment by nevack** [comment]
Removed `"STATIC=true"`, made `gcc` a hard dependency.

File: Formula/btop.rb [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  homepage "https://github.com/aristocratos/btop"
+  url "https://github.com/aristocratos/btop/archive/refs/tags/v1.1.2.tar.gz"
+  sha256 "997376596cc6ca808b5a26f9fe25da274c19c2970aaf757943a009f4c28b758c"
+  license "Apache-2.0"
+
+  depends_on "coreutils" => :build
+
+  on_macos do
+    depends_on "gcc"
+  end
```
Comment (carlocab) [inline Formula/btop.rb:12]: You need this on Linux now too.

**Review by carlocab — APPROVED** [review]
Thanks again, @nevack!

**Comment by aristocratos** [comment]
@nevack @carlocab 
Hi, thanks for adding a formula for btop :)

Clang is not going to be able to compile btop until it's support for C++20 is on par with GCC. C++20 Ranges is currently what's  holding it back. 

> [...] but it appears this will statically link glibc on Linux, which is something we're unlikely to accept.

The release binaries for linux are statically linked with musl and makes the binaries compatibly with _any_ linux system running kernel 2.6.39 or newer.
The `STATIC=true` flag is however never enabled by default (or recommended) when manually compiling for any platform.

**Comment by nevack** [comment]
> @nevack @carlocab Hi, thanks for adding a formula for btop :)

Happy to help, sorry for not informing you about this formula added.

> The `STATIC=true` flag is however never enabled by default (or recommended) when manually compiling for any platform.

Sorry for misinforming, I thought it's always recommended to use this option, as all public releases are compiled with it.

**Pull request closed without merging**